specification/dartLangSpec.tex

```diff
@@ -16347,19 +16389,32 @@ \subsubsection{The Imported Namespace}
 
 \LMHash{}%
 Let $i \in 1 .. m$.
+
+\LMHash{}%
+{\bf Step one.}
 In the first step we compute the namespace provided by each imported library,
+respectively by each group of libraries imported with the same prefix,
 \NamespaceName{\metavar{one},i}.
 
 \LMHash{}%
 \Case{Step one for imports without a prefix}
 When $I_i$ has no prefix, \NamespaceName{\metavar{one},i} is
 the namespace obtained from applying the namespace combinators of $I_i$ to
 the exported namespace of $L_i$
-(\ref{namespaceCombinators}).
+(\ref{namespaceCombinators}),
+and eliminating every binding
+where the key is an import prefix declared in $L$.
```

Comment:
Why treat import prefixes specially here? 

We need to ensure that library members shadows imports, so why have two different ways to deal with library members and import prefixes. 
Can we not just treat import prefixes the same way as declared names in every way (we already stated above that they conflict with member declarations).

Reply:
We used to say that the local namespace NS_local of the current library contains bindings for import prefixes. But that was incorrect, because there was no reasonable way to specify what each import prefix is bound to (that's the purpose of this whole section, which uses NS_local).

Also, it is allowed to have several imports with the same prefix, which means that we cannot just say that each import directive with prefix `p` is (with respect to namespaces) treated just like a declaration named `p`.

In order to ensure that the conflict merging step doesn't have a conflict at a prefix name (which would make that name conflicted, which is wrong: the prefix must prevail), I added the step here that eliminates any name clashes between prefixes and names that are imported with no prefix.

The rule that makes it an error to have a declaration and a prefix with the same basename is needed in order to avoid the (well-defined, but useless) shadowing of a prefix by a local declaration. This _does_ make prefixes even more similar to other declared names, but I still think that there are so many differences that we need to maintain the distinction.

Reply:
No matter how we wrangle it, each distinct prefix, `p`, introduced by `as p` on one or more imports, gives rise to a binding from `p` to a namespace (let's call it `NS_p`). 
The only real question is where that binding lives. Which bindings `NS_p` has is something we'll have to figure out independently of where `p` lives.

The *behavior* here is that:
* A prefix `p` conflicts with a library member declaration named `p`, and
* A prefix `p` takes precedence over an imported declaration named `p`.

That still sounds like `p` should introduce an entry in the local namespace for `p`, because it behaves *exactly* as if it did.
I can't see why it shouldn't be possible.

Reply:
Surely it's not impossible. Currently the namespace NS_local contains local declarations, and prefixes are contained in NS_import, and then we obtain the library namespace as a union including those two.

But do you think it actually matters whether those bindings are included in one or the other intermediate result?

The point is that _'exactly'_ isn't accurate: We can't just say that each occurrence of `as p` is treated like a declaration in the current library because there can be more than one, and we can't say what `p` maps to without talking about each library, and acting on the case where it is imported with a prefix.

Those issues are handled with the current approach, because each prefixed import gives rise to the mapping of `p` for the whole group prefixed by `p`, and the union of these bindings is trivial because all members of that group have exactly the same binding.

So we could just say that `NS_import` is the conflict merging among the `NS_one,i` for such `i` that the import has no prefix, and then we compute another intermediate result `NS_local U NS_one,j1 .. U NS_one,jk` where `j1 .. jk` are all the imports with a prefix. 

But I'm not convinced that it's going to improve the readability: We would then _not_ provide access to all imported names in `NS_import`, which seems to be a similarly confusing outcome.